### PR TITLE
[SPARK-31391][SQL][TEST] Add AdaptiveTestUtils to ease the test of AQE

### PR DESCRIPTION
**Pull request by Ngone51**
### What changes were proposed in this pull request?

This PR adds `AdaptiveTestUtils` to make AQE test simpler, which includes:

`DisableAdaptiveExecution` - a test tag to skip a single test case if AQE is enabled.
`EnableAdaptiveExecutionSuite` - a helper trait to enable AQE for all tests except those tagged with `DisableAdaptiveExecution`.
`DisableAdaptiveExecutionSuite` - a helper trait to disable AQE for all tests.
`assertExceptionMessage` - a method to handle message of normal or AQE exception in a consistent way.
`assertExceptionCause` - a method to handle cause of normal or AQE exception in a consistent way.

### Why are the changes needed?

With this utils, we can:
- reduce much more duplicate codes;
- handle normal or AQE exception in a consistent way;
- improve the stability of AQE tests;

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Updated tests with the util.

### CODE REVIEW

**Comment by Ngone51** [comment]
cc @cloud-fan @maryannxue

File: sql/core/src/test/scala/org/apache/spark/sql/MetadataCacheSuite.scala [hunk 1]
```diff
@@ -55,8 +56,8 @@ abstract class MetadataCacheSuite extends QueryTest with SharedSparkSession {
       val e = intercept[SparkException] {
         df.count()
       }
-      assert(e.getMessage.contains("FileNotFoundException"))
-      assert(e.getMessage.contains("recreating the Dataset/DataFrame involved"))
+      assertExceptionMessage(e, "FileNotFoundException")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/MetadataCacheSuite.scala:59]: `assertErrorMsgWithAQE`?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala [hunk 2]
```diff
@@ -26,29 +26,18 @@ import org.apache.spark.sql._
...
 import org.apache.spark.sql.test.SharedSparkSession
 import org.apache.spark.util.{AccumulatorContext, JsonProtocol}
 
-class SQLMetricsSuite extends SharedSparkSession with SQLMetricsTestUtils {
+class SQLMetricsSuite extends SharedSparkSession with SQLMetricsTestUtils
+  with DisableAdaptiveExecution {
   import testImplicits._
 
-  var originalValue: String = _
-  // With AQE on/off, the metric info is different.
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala:40]: can we keep the comment somewhere?

File: sql/core/src/test/scala/org/apache/spark/sql/MetadataCacheSuite.scala (hunk 1 above)
Comment (Ngone51) [reply sql/core/src/test/scala/org/apache/spark/sql/MetadataCacheSuite.scala:59]: The helper method can also work for normal cases without AQE but adding `AQE` just sounds like it's only used by AQE?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/WholeStageCodegenSuite.scala [hunk 3]
```diff
@@ -28,23 +29,12 @@ import org.apache.spark.sql.internal.SQLConf
 import org.apache.spark.sql.test.SharedSparkSession
 import org.apache.spark.sql.types.{IntegerType, StringType, StructType}
 
-class WholeStageCodegenSuite extends QueryTest with SharedSparkSession {
+// Disable AQE because the WholeStageCodegenExec rule is applied when running QueryStageExec
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/WholeStageCodegenSuite.scala:32]: `is applied` -> `is added`. There are more mistakes like this, please fix all of them

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveTestUtils.scala [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+import java.io.{PrintWriter, StringWriter}
+
+import org.scalactic.source.Position
+import org.scalatest.Tag
+
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.test.SQLTestUtils
+
+/**
+ * Don't run this test at all if adaptive execution is enabled.
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveTestUtils.scala:29]: `Make sure the test is run with adaptive execution disabled.`

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master/3.0!

**Comment by Ngone51** [comment]
thank you!

**Pull request closed without merging**